extensions/smb/SmbConnectionControllerService.h

```diff
@@ -0,0 +1,101 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <filesystem>
+#include <string>
+#include <memory>
+
+#include "ProcessContext.h"
+#include "core/PropertyDefinition.h"
+#include "core/PropertyDefinitionBuilder.h"
+#include "core/controller/ControllerService.h"
+#include "core/logging/Logger.h"
+#include "core/logging/LoggerConfiguration.h"
+#include "utils/Enum.h"
+#include "utils/expected.h"
+
+namespace org::apache::nifi::minifi::extensions::smb {
+
+class SmbConnectionControllerService : public core::controller::ControllerService {
+ public:
+  EXTENSIONAPI static constexpr const char* Description = "SMB Connection Controller Service";
+
+  EXTENSIONAPI static constexpr auto Hostname  = core::PropertyDefinitionBuilder<>::createProperty("Hostname")
+      .withDescription("The network host to which files should be written.")
+      .isRequired(true)
+      .build();
+  EXTENSIONAPI static constexpr auto Share = core::PropertyDefinitionBuilder<>::createProperty("Share")
+      .withDescription(R"(The network share to which files should be written. This is the "first folder" after the hostname: \\hostname\[share]\dir1\dir2)")
+      .isRequired(true)
+      .build();
+  EXTENSIONAPI static constexpr auto Domain  = core::PropertyDefinitionBuilder<>::createProperty("Domain")
+      .withDescription("The domain used for authentication. Optional, in most cases username and password is sufficient.")
+      .isRequired(false)
+      .build();
+  EXTENSIONAPI static constexpr auto Username  = core::PropertyDefinitionBuilder<>::createProperty("Username")
+      .withDescription("The username used for authentication. If no username is set then anonymous authentication is attempted.")
+      .isRequired(false)
+      .build();
+  EXTENSIONAPI static constexpr auto Password  = core::PropertyDefinitionBuilder<>::createProperty("Password")
+      .withDescription("The password used for authentication. Required if Username is set.")
+      .isRequired(false)
+      .build();
```

Comment:
can we mark these with `withDependentProperties` of each other?

Reply:
good idea, added it in https://github.com/apache/nifi-minifi-cpp/pull/1634/commits/3b513dcd2120d5e5ab361e8230060721450c6551#diff-5b6eb650aaa403ae0dd4e89f028e9d80385743d7d67515250957b6464887ae03R53-R58